### ISSUE TITLE
Debug Assert error in MethodToStateMachineRewriter

### ISSUE THREAD

**Issue by bernd5** [dotnet/roslyn#74362]
**Version Used**: 

**Steps to Reproduce**:
Compile the following code with debug asserts enabled:

```csharp
using System;
using System.Collections.Generic;
using System.Threading.Tasks;

public class App
{
    public static void Main()
    {
        foreach (var x in new X().ValuesViaLocalFunction)
        {
            Console.WriteLine(x);
        }
    }
}


public class X
{
    public async IAsyncEnumerable<int> GetValues(bool extra = true)
    {
       yield return 0;
    }

    public IEnumerable<int> ValuesViaLocalFunction
    {
        get
        {
            foreach (var b in Do(true).ToBlockingEnumerable())
            {
                yield return b;
            }

            async IAsyncEnumerable<int> Do(bool extra)
            {
                await foreach (var v in GetValues(true))
                {
                    yield return v;
                }
            }
        }
    }
}
```

**Expected Behavior**:
No assert triggered.

**Actual Behavior**:
Debug assert is triggered at: [Link](https://github.com/dotnet/roslyn/blob/7241b8522990905c491dee920d32dc09c345920d/src/Compilers/CSharp/Portable/Lowering/StateMachineRewriter/MethodToStateMachineRewriter.cs#L220)

The reason is that `AsyncExceptionHandlerRewriter` does not care about `SyntheticBoundNodeFactory.Syntax`. The syntax needs to be updated to the try statement.
See: [Link](https://github.com/dotnet/roslyn/blob/7241b8522990905c491dee920d32dc09c345920d/src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncExceptionHandlerRewriter.cs#L183)

**Comment by AlekseyTs** [comment]
Assigning back to Julien since he already has a PR with the fix
